src/scripts/underground/Mine.ts

```diff
@@ -33,29 +35,37 @@ class Mine {
         Mine.grid = tmpGrid;
         Mine.rewardGrid = tmpRewardGrid;
 
-        let added = 0;
-        for (let i = 0; i < App.game.underground.getMaxItems(); i++) {
-            const item = UndergroundItems.getRandomItem();
-            const x = Mine.getRandomCoord(Underground.sizeX, item.space[0].length);
-            const y = Mine.getRandomCoord(App.game.underground.getSizeY(), item.space.length);
-            const res = Mine.canAddReward(x, y, item);
-            if (res) {
-                Mine.addReward(x, y, item);
-                added = added + 1;
-            }
+        // Generate items for new layer
+        // Number of underground items must be >= min
+        // If max > min, randomly select the number of items for the layer
+        // Otherwise, the number of items must be equivalent to min
+        let numItems = App.game.underground.getMinItems();
+        if (App.game.underground.getMinItems() < App.game.underground.getMaxItems()) {
+            numItems = Rand.intBetween(App.game.underground.getMinItems(), App.game.underground.getMaxItems());
         }
 
-        // Check in case player upgrade min above max
-        const min = Math.min(App.game.underground.getMinItems(), App.game.underground.getMaxItems());
-        while (added < min) {
+        // Add numItems items to the layer
+        for (let i = 0; i < numItems; i++) {
+            let res = false;
+            let x = 0;
+            let y = 0;
             const item = UndergroundItems.getRandomItem();
-            const x = Mine.getRandomCoord(Underground.sizeX, item.space[0].length);
-            const y = Mine.getRandomCoord(App.game.underground.getSizeY(), item.space.length);
-            const res = Mine.canAddReward(x, y, item);
+            let attempts = 0;
+            // Keep checking random spots until a legal spot is found
+            // If too many failed attempts are made, break out as a failsafe
+            while (!res && attempts++ < this.maxPlacementAttempts) {
+                this.rotateReward(item);
+                x = Mine.getRandomCoord(Underground.sizeX, item.space[0].length);
+                y = Mine.getRandomCoord(App.game.underground.getSizeY(), item.space.length);
+                res = Mine.canAddReward(x, y, item);
+            }
+            // If item can be added, add it
+            // Else, if we haven't hit the minimum number of items yet, try again with a new random item
+            // Otherwise, just skip to next item
             if (res) {
                 Mine.addReward(x, y, item);
-                added = added + 1;
-                //This should loop until it's added.
+            } else if (i < App.game.underground.getMinItems()) {
+                i--;
```

Comment:
Just as a note, this could be changed to just a normal `else` statement if we wanted this to be applied to all items and not just the minimum required ones.

Reply:
This scares me a little because it looks like we can get stuck in an infinite loop if no item can be added anywhere on the grid.

But we are already doing an infinite loop until `min` and if we remove it, it's possible to get layers with less items than we are promising with the min items upgrade...

Problem for another day I think.